### ISSUE TITLE
How to calculate chromatin accessibility for non ATAC-seq

### ISSUE THREAD

**Issue by JingGuo1997** [snehamitra/SCARlink#23]
Hi @snehamitra ，

Thank you for developing SCARlink — it's a powerful and promising tool for analyzing potential enhancers.

I’m currently trying to apply SCARlink to my dataset, but I’ve run into a challenge. Unlike the typical scATAC-seq input, I only have a chromatin accessibility table (rather than raw scATAC-seq data), and therefore I’m unable to construct an ArchR object, which SCARlink requires.

Here is an example of the accessibility bedgraph I’m working with:

> chrom   start   end     20211020-YR-1-1_sc10    20211020-YR-1-1_sc1     20211020-YR-1-1_sc28    20211020-YR-1-1_sc42       20211020-YR-1-1_sc47    20211020-YR-1-1_sc57    20211020-YR-1-1_sc5
chr1    0       500     .       .       .       .       .       .       .
chr1    500     1000    .       .       .       .       .       .       .
chr1    1000    1500    .       .       .       .       .       .       0
chr1    1500    2000    .       .       .       .       .       .       .
chr1    2000    2500    .       .       .       .       .       .       .
chr1    2500    3000    .       .       .       .       .       .       .
chr1    3000    3500    .       .       .       .       .       .       .
chr1    3500    4000    .       .       .       .       .       .       1
chr1    4000    4500    .       .       .       .       .       .       0.25

In addition, I also have a scRNA-seq dataset available in CSV format:
>20211020-YR-1-2_sc28    20211020-YR-1-1_sc1     20211020-YR-1-1_sc42    20211020-YR-1-1_sc57    20211020-YR-1-1_sc66
A1CF    0       0       0       0
A2M     0       0.388800274154032       0       0
A2ML1   0       0       0       0
A3GALT2 0       0       0       0

So I would like to directly use the core functions in your source code scarlink/src/model.py. However, I'm having difficulty understanding exactly what kind of input your model requires. 

Would it be possible for you to help me understand how to construct the required input .h5 file for the model using my data?

I would greatly appreciate any guidance you can provide.

Estelle GUO，06/05/2025

**Comment by JingGuo1997** [comment]
> Hi [@snehamitra](https://github.com/snehamitra) ，
> 
> Thank you for developing SCARlink — it's a powerful and promising tool for analyzing potential enhancers.
> 
> I’m currently trying to apply SCARlink to my dataset, but I’ve run into a challenge. Unlike the typical scATAC-seq input, I only have a chromatin accessibility table (rather than raw scATAC-seq data), and therefore I’m unable to construct an ArchR object, which SCARlink requires.
> 
> Here is an example of the accessibility bedgraph I’m working with:
> 
> > chrom   start   end     20211020-YR-1-1_sc10    20211020-YR-1-1_sc1     20211020-YR-1-1_sc28    20211020-YR-1-1_sc42       20211020-YR-1-1_sc47    20211020-YR-1-1_sc57    20211020-YR-1-1_sc5
> > chr1    0       500     .       .       .       .       .       .       .
> > chr1    500     1000    .       .       .       .       .       .       .
> > chr1    1000    1500    .       .       .       .       .       .       0
> > chr1    1500    2000    .       .       .       .       .       .       .
> > chr1    2000    2500    .       .       .       .       .       .       .
> > chr1    2500    3000    .       .       .       .       .       .       .
> > chr1    3000    3500    .       .       .       .       .       .       .
> > chr1    3500    4000    .       .       .       .       .       .       1
> > chr1    4000    4500    .       .       .       .       .       .       0.25
> 
> In addition, I also have a scRNA-seq dataset available in CSV format:
> 
> > 20211020-YR-1-2_sc28    20211020-YR-1-1_sc1     20211020-YR-1-1_sc42    20211020-YR-1-1_sc57    20211020-YR-1-1_sc66
> > A1CF    0       0       0       0
> > A2M     0       0.388800274154032       0       0
> > A2ML1   0       0       0       0
> > A3GALT2 0       0       0       0
> 
> So I would like to directly use the core functions in your source code scarlink/src/model.py. However, I'm having difficulty understanding exactly what kind of input your model requires.
> 
> Would it be possible for you to help me understand how to construct the required input .h5 file for the model using my data?
> 
> I would greatly appreciate any guidance you can provide.
> 
> Estelle GUO，06/05/2025

**Comment by JingGuo1997** [comment]
I solved this problem. Here is my script！
```
import h5py
import sys
import pandas as pd
import numpy as np
from scipy import sparse

gene=sys.argv[1]
# 1. 读取基因表达数据
rna_exp = pd.read_csv('rna_exp.csv', 
                      index_col=0,      # 第一列作为索引
                      sep='\t',         # 使用制表符作为分隔符
                      na_values=['NA', 'nan', 'NaN'],  # 指定缺失值
                      dtype={'gene': str}  # 指定列的数据类型
                      )
# 2. 读取细胞信息
cell_info = pd.read_csv('cell_info.txt', sep='\t')
cell_info['ReadsInTSS'] = 1
print(cell_info.head())
print(cell_info.shape)
# 检查细胞名是否一致
assert all(rna_exp.columns == cell_info["cell_name"]), "表达矩阵和细胞信息的细胞顺序不一致！"
# 3、读取 tiles 文件，header=0 表示第一行为表头
tiles = pd.read_csv(
    f'genes_CA_input/{gene}_window_intersect.bedgraph',
    sep='\t',
    header=0,
    na_values='.'
)

# 用0填充所有NA
tiles = tiles.fillna(0)

# 检查前几行
print(tiles.head())
print(tiles.shape)
print(tiles.iloc[:5, :8])  # 查看前5行前8列
# tile 区域信息
tile_info = tiles.iloc[:, :3]  # chr, start, end

# 可及性矩阵（行为tile，列为细胞）
tile_matrix = tiles.iloc[:, 3:]

# 检查列名是否和表达矩阵一致
print(tile_matrix.columns[:5])
print(rna_exp.columns[:5])
# 只保留表达矩阵中有的细胞，并按顺序排列
tile_matrix_aligned = tile_matrix[rna_exp.columns]
print((tile_matrix_aligned.columns == rna_exp.columns).all())  # 应该输出 True
tile_matrix_aligned=tile_matrix_aligned.transpose()
# 1. tile矩阵转为csr稀疏矩阵（行为tile，列为细胞）
tile_csr = sparse.csr_matrix(tile_matrix_aligned.values)

# 2. 基因表达矩阵转为csr稀疏矩阵（行为基因，列为细胞）
rna_csr = sparse.csr_matrix(rna_exp.values)
print("稀疏矩阵形状:", rna_csr.shape)
print("非零元素数量:", rna_csr.nnz)
tile_info = tile_info.rename(columns={'chrom': 'seqnames', 'start': 'start', 'end': 'end'})
with h5py.File('coassay_matrix.h5', 'w') as f:
    # --- 写入tile矩阵 ---
    grp = f.create_group(gene)
    grp.create_dataset('data', data=tile_csr.data)
    grp.create_dataset('indices', data=tile_csr.indices)
    grp.create_dataset('indptr', data=tile_csr.indptr)
    grp.create_dataset('shape', data=tile_csr.shape)
    # --- 写入gene_expression ---
    gex_grp = f.create_group('gene_expression')
    gex_grp.create_dataset('data', data=rna_csr.data)
    gex_grp.create_dataset('indices', data=rna_csr.indices)
    gex_grp.create_dataset('indptr', data=rna_csr.indptr)
    gex_grp.create_dataset('shape', data=rna_csr.shape)
    gex_grp.create_dataset('gene_info', data=np.array(rna_exp.index).astype('S'))
# tile_info
tile_info.to_hdf('coassay_matrix.h5', key=f'{gene}/tile_info', format='table')
# gene_info为基因名
gene_info_df = pd.DataFrame({'gene_name': rna_exp.index})
gene_info_df.to_hdf('coassay_matrix.h5', key='gene_expression/gene_info', format='table')     
gene_info_df['gene_name'] == gene
print(gene in gene_info_df['gene_name'].values)
#cell_info
cell_info.to_hdf('coassay_matrix.h5', key='cell_info', format='table')                 
```
and

```
import sys
sys.path.append('/data/data_guojing/enhancer_mf_240405/scarlink/SCARlink')
from scarlink.src.model import RegressionModel
import matplotlib
import matplotlib.pyplot as plt
import matplotlib.colorbar as colorbar
import matplotlib.gridspec as gridspec
import matplotlib.colors as colors
from matplotlib import cm
import logging
import seaborn
import tensorflow as tf
import numpy as np
import pandas
import h5py
import sys
import os
import warnings
from tables import NaturalNameWarning
from scipy import stats
from scipy.sparse import csr_matrix
from sklearn.preprocessing import MaxAbsScaler
from sklearn.model_selection import RepeatedKFold, train_test_split
import tensorflow.keras.backend as K
from scarlink.src.plotExtra import plotRegion, get_fragment_counts, plot_hist, create_colormap
from scarlink.src.read_h5_and_group_cells import construct_cell_info, construct_gex_mat, get_train_test_split, get_gene_tile_matrix_group_cells, write_significance, read_sparse_significance
from scarlink.src.tile_significance import set_gene_tile_significance_bootstrapped, set_gene_tile_significance_signed_rank

gene = sys.argv[1]
rm = RegressionModel(
    input_file='coassay_matrix.h5',      # 你的h5输入文件
    output_dir='scarlink_out/',          # 输出目录（自动创建）
    log_file_name='scarlink.log',        # 日志文件
    gtf_file='/data/data_guojing/mf_ChaRM-seq_0407/refGene.gtf.gz'                  # 基因组注释文件（mf5.gtf）
)
#训练20次选结果最好的
rm.train_test_model(
    gene=gene,
    normalization_factor='ReadsInTSS',  # 如果你用的是全1归一化因子，或你cell_info里有别的归一化列就填那个
    max_zero_fraction=0.9,         # 稀疏度阈值，默认即可
    epochs=20,                     # 训练轮数，可根据数据量调整
    verbose=True,
    plot_loss=True,
    
)
w_mat, intercept = rm.get_gene_coefficient(gene)
print("回归系数形状：", w_mat.shape)
print("截距：", intercept)
#读取h5并写成bedgraph
corbed=h5py.File("scarlink_out/coefficients.hd5", mode = 'r')
chrom = corbed['genes/' + gene].attrs['chr']
start = corbed['genes/' + gene].attrs['start']
end = corbed['genes/' + gene].attrs['end']
values =corbed['genes/' + gene][:]
tilesize=500
chrom,start,end,values,tilesize
with open(f"scarlink_out/{gene}_coefficients.bedgraph", "w") as f:
    for i, v in enumerate(values):
        tile_start = start + i * tilesize
        tile_end = tile_start + tilesize
        f.write(f"{chrom}\t{tile_start}\t{tile_end}\t{float(v)}\n")
#shap 值
celltype_col = 'celltype'
zscore_d = rm.compute_gene_tile_significance_shap(gene, celltype_col)
cell_types = ['EPI', 'PS', 'Ecto', 'Meso', 'Endo']
length = len(zscore_d[cell_types[0]])  # 假设每个array长度一致
with open(f"scarlink_out/{gene}_shapely.bedgraph", "w") as f:
    # 可选：写表头
    f.write("chr\tstart\tend\t" + "\t".join(cell_types) + "\n")
    for i in range(length):
        tile_start = start + i * tilesize
        tile_end = tile_start + tilesize
        values = [float(zscore_d[ct][i]) for ct in cell_types]
        line = f"{chrom}\t{tile_start}\t{tile_end}\t" + "\t".join(map(str, values)) + "\n"
        f.write(line)
rm.plot_gene(
    gene=gene,plot_shap=True,
    groups='celltype',         # cell_info中分组列名
    to_save=True,
    plot_dir='scarlink_out/plots/',
    cluster_order=['EPI', 'PS', 'Ecto', 'Meso', 'Endo'] ,
    save_format='pdf'          # 指定保存为pdf矢量图
)

```

**Comment by snehamitra** [comment]
Thanks for sharing your script! One thing to add would be the normalization factor for scATAC-seq. In your script, ReadsInTSS is set to 1. Instead, you can set nFrags to tiles.sum(1). I updated the preprocessing step so now you can take anndata as input. So you can generate an anndata object using your tile matrix and then follow the rest of the pipeline. You can find more preprocessing details in this [notebook](https://github.com/snehamitra/SCARlink_private/blob/main/notebooks/preprocessing_scRNA_scATAC_python.ipynb).

**Comment by JingGuo1997** [comment]
Thank you for your help！ I will revise my script in the subsequent analysis.
